portable: add 'reattach' verb and DBUS interface

bluca opened the pull request:
Add 'upgrade' verb to portablectl, and corresponding DBUS interface
to systemd-portabled.
Takes the same parameters as 'attach', but it will do a 'detach' (and
it will refuse to proceed if it cannot be done) first, matching on
the unversioned prefix of the new image. Eg:

portablectl upgrade /tmp/foo_2.raw

will cause foo_1.raw to be detached.

The key difference with a manual 'detach old' plus 'attach new' is that
the running units are not disturbed until after the attach completed,
and if --now is passed they are then restarted.
A 'detach' is not allowed normally if the units are running.

By using a restart-after-deploy method, 'upgrade' allows for minimal
interruption of service and also for features that only work on restart
(eg: file descriptor store) to work as intended.

Fixes https://github.com/systemd/systemd/issues/13621

poettering commented:
hmm, not sure i grok this. what happens if the unit files in the image change, or one is added or removed?

bluca commented:
> hmm, not sure i grok this. what happens if the unit files in the image change, or one is added or removed?

If it changes or something is added, then it all works as expected: daemon-reload after the UpgradeImage returns, and a Restart is called on all the units listed as "changes" in the return value of the dbus method, just like on Attach.

I did forget to handle the case when a unit is removed though (d'oh) as it's not my typical use case - the return list can be used to handle that as well, I'll fix it tomorrow so that units that disappear are stopped instead.

poettering commented:
btw, why call this "upgrade"? why not just "reattach"? i.e. is there anything that implies upwards ordering of the version?

poettering commented on src/portable/portablectl.c:
```diff
@@ -926,6 +1020,9 @@ static int help(int argc, char *argv[], void *userdata) {
                "  set-limit [NAME|PATH]       Set image or pool size limit (disk quota)\n"
+               "  upgrade NAME|PATH [PREFIX...]\n"
+               "                              Upgrades and restarts the specified portable\n"
+               "                              service image\n"
```
i'd move this up to where attach/detach is listed, it has nmore to to with that.

(this is ordered alphabetically only by accident, we don't actually intend to order things alphabetically in --help texts, but group things by context)

bluca replied:
moved

poettering commented:
hmm, and what about enabling units when oyu do an upgrade like that? restarting is one thing, but there might be new units to enable

bluca commented:
> btw, why call this "upgrade"? why not just "reattach"? i.e. is there anything that implies upwards ordering of the version?

No particular reason - I mean my use case is having fast servicing when I need to upgrade the image (CVE fixes or whatever), so that's where it came from. I'll change it to reattach.

> hmm, and what about enabling units when oyu do an upgrade like that? restarting is one thing, but there might be new units to enable

Good point, will fix that as well tomorrow.

bluca commented:
Renamed the verb/interface from 'upgrade' to 'reattach' as requested, and also now handling the case where units are removed or are new and need to be enabled. PTAL!

poettering commented on man/portablectl.xml:
```diff
@@ -155,6 +155,20 @@
+        <term><command>reattach</command> <replaceable>IMAGE</replaceable> [<replaceable>PREFIX…</replaceable>]</term>
+
+        <listitem><para>Detaches an existing portable service image from the host, and immediately attaches the new
+        specified version. Running units are not stopped during the process. Partial matching, to allow for different
```
"new specified version"? there's no "version" specified here.

Just maybe say "Detaches an existing portable service image from the host, and immediately attaches it again. This is useful in case the image was replaced" or so

bluca replied:
Leftover from when the api was called 'upgrade' - changed as suggested. The idea btw as you saw later is that matching is relaxed - so doing 'reattach image_0.2.raw' will match an existing attached image_0.1.raw

poettering commented on man/portablectl.xml:
```diff
@@ -155,6 +155,20 @@
+        specified version. Running units are not stopped during the process. Partial matching, to allow for different
+        versions in the image name, is allowed: only the part before the first <literal>_</literal> character have
+        to match. If the new image doesn't exist, the existing one will not be detached. The parameters follow the
+        same syntax as the <command>attach</command> command.</para></listitem>
```
hmm, the prefix handling is the same as for "attach", no? there are no different semantics here? if there are, why? if not, maybe just reference the documentation for "attach" above

bluca replied:
Yes it's all the same. Shortened to just mention the 'attach' case.

poettering commented on src/portable/portable.c:
```diff
@@ -1026,7 +1026,7 @@ int portable_attach(
 }
 
-static bool marker_matches_image(const char *marker, const char *name_or_path) {
+static bool marker_matches_image(const char *marker, const char *name_or_path, bool unversioned_match) {
```
hmm, is there a good reason to do version exact matches here ever? maybe e should just losen this up, and always make the version-ignorant checks here?

poettering replied:
at least i don#t see a reason why we shoud be so strict here.

bluca replied:
I'm not sure - existing behaviour, mostly. Maybe we want 'detach foo_1.2.raw' to really remove just that? I'll loosen it for now.

poettering commented on src/portable/portablectl.c:
```diff
@@ -508,7 +510,82 @@ static int maybe_enable_start(sd_bus *bus, sd_bus_message *reply) {
+
+                if (streq(type, "unlink") && ENDSWITH_SET(path, ".service", ".target", ".socket")) {
+                        (void) maybe_start_stop_restart(bus, path, "StopUnit", wait);
+                }
```
coding style, no {} around single line if blocks

poettering commented on src/portable/portablectl.c:
```diff
@@ -508,7 +510,82 @@ static int maybe_enable_start(sd_bus *bus, sd_bus_message *reply) {
 
                 if (STR_IN_SET(type, "symlink", "copy") && ENDSWITH_SET(path, ".service", ".target", ".socket")) {
```
hmm, the unit type list seems out of date, .path and .timer is missing there, iiuc

bluca replied:
added and refactored

poettering commented on src/portable/portablectl.c:
```diff
@@ -508,7 +510,82 @@ static int maybe_enable_start(sd_bus *bus, sd_bus_message *reply) {
+                if (r == 0)
+                        break;
+
+                if (STR_IN_SET(type, "symlink", "copy") && ENDSWITH_SET(path, ".service", ".target", ".socket")) {
```
hmm, these lists are repeated too often for my taste, and given they already got out of date, should be unified in one helper call:

```c
static bool is_portable_managed(const char *unit) {
        return ENDSWITH_SET(…
```

or so

bluca replied:
sure, done

poettering commented on src/portable/portabled-image-bus.c:
```diff
@@ -422,6 +422,190 @@ static int bus_image_method_remove(sd_bus_message *message, void *userdata, sd_b
+                goto finish;
+
+        /* We want to return the list of units really removed by the detach,
+         * and not added again by the attach */
```
is this really so? i mean, if the unit files changed we might want to restart the unit too, no?

bluca replied:
Yes - that's the second list in the return values signature. You need to be able to tell which is which, so they are separated.

bluca replied:
From the commit message:

> The DBUS interface returns two lists: first the removals from the detach
that were not immediately re-added in the attach, so that the caller
can stop the relevant units, and then the list of additions that are
either new or updates, so that the caller can restart/enable the
relevant units. portablectl already implements this with the existing
--now/--enable switches.

bluca commented:
Updated again. Also moved the mksquashfs outside of the test script - as the mksquashfs command segfaults on the ubuntu CI (!!?)

bluca commented:
Here's an example attach -> reattach -> reattach -> detach that should exemplify how it works in practice. Two directories, minimal_1 and minmal_2, have minimal.service in common. One has minimal-foo.service and the other has minimal-bar.service.

```
$ sudo portablectl attach --runtime --now /tmp/minimal_1 minimal
(Matching unit files with prefixes 'minimal'.)
Created directory /run/systemd/system.attached.
Created directory /run/systemd/system.attached/minimal-foo.service.d.
Written /run/systemd/system.attached/minimal-foo.service.d/20-portable.conf.
Created symlink /run/systemd/system.attached/minimal-foo.service.d/10-profile.conf → /usr/lib/systemd/portable/profile/default/service.conf.
Copied /tmp/minimal_1/usr/lib/systemd/system/minimal-foo.service → /run/systemd/system.attached/minimal-foo.service.
Created directory /run/systemd/system.attached/minimal.service.d.
Written /run/systemd/system.attached/minimal.service.d/20-portable.conf.
Created symlink /run/systemd/system.attached/minimal.service.d/10-profile.conf → /usr/lib/systemd/portable/profile/default/service.conf.
Copied /tmp/minimal_1/usr/lib/systemd/system/minimal.service → /run/systemd/system.attached/minimal.service.
Created symlink /run/portables/minimal_1 → /tmp/minimal_1.
Queued /org/freedesktop/systemd1/job/19276 to call StartUnit on portable service minimal-foo.service.
Queued /org/freedesktop/systemd1/job/19359 to call StartUnit on portable service minimal.service.
```

```
$ sudo portablectl reattach --runtime --now /tmp/minimal_2 minimal
(Matching unit files with prefixes 'minimal'.)
Removed /run/systemd/system.attached/minimal-foo.service.
Removed /run/systemd/system.attached/minimal-foo.service.d.
Removed /run/portables/minimal_1.
Created directory /run/systemd/system.attached.
Created directory /run/systemd/system.attached/minimal.service.d.
Written /run/systemd/system.attached/minimal.service.d/20-portable.conf.
Created symlink /run/systemd/system.attached/minimal.service.d/10-profile.conf → /usr/lib/systemd/portable/profile/default/service.conf.
Copied /tmp/minimal_2/usr/lib/systemd/system/minimal.service → /run/systemd/system.attached/minimal.service.
Created directory /run/systemd/system.attached/minimal-bar.service.d.
Written /run/systemd/system.attached/minimal-bar.service.d/20-portable.conf.
Created symlink /run/systemd/system.attached/minimal-bar.service.d/10-profile.conf → /usr/lib/systemd/portable/profile/default/service.conf.
Copied /tmp/minimal_2/usr/lib/systemd/system/minimal-bar.service → /run/systemd/system.attached/minimal-bar.service.
Created symlink /run/portables/minimal_2 → /tmp/minimal_2.
Queued /org/freedesktop/systemd1/job/19442 to call StopUnit on portable service minimal-foo.service.
Queued /org/freedesktop/systemd1/job/19443 to call RestartUnit on portable service minimal.service.
Queued /org/freedesktop/systemd1/job/19526 to call RestartUnit on portable service minimal-bar.service.
```

```
$ sudo portablectl reattach --runtime --now /tmp/minimal_1 minimal
(Matching unit files with prefixes 'minimal'.)
Removed /run/systemd/system.attached/minimal-bar.service.
Removed /run/systemd/system.attached/minimal-bar.service.d.
Removed /run/portables/minimal_2.
Created directory /run/systemd/system.attached.
Created directory /run/systemd/system.attached/minimal-foo.service.d.
Written /run/systemd/system.attached/minimal-foo.service.d/20-portable.conf.
Created symlink /run/systemd/system.attached/minimal-foo.service.d/10-profile.conf → /usr/lib/systemd/portable/profile/default/service.conf.
Copied /tmp/minimal_1/usr/lib/systemd/system/minimal-foo.service → /run/systemd/system.attached/minimal-foo.service.
Created directory /run/systemd/system.attached/minimal.service.d.
Written /run/systemd/system.attached/minimal.service.d/20-portable.conf.
Created symlink /run/systemd/system.attached/minimal.service.d/10-profile.conf → /usr/lib/systemd/portable/profile/default/service.conf.
Copied /tmp/minimal_1/usr/lib/systemd/system/minimal.service → /run/systemd/system.attached/minimal.service.
Created symlink /run/portables/minimal_1 → /tmp/minimal_1.
Queued /org/freedesktop/systemd1/job/19609 to call StopUnit on portable service minimal-bar.service.
Queued /org/freedesktop/systemd1/job/19610 to call RestartUnit on portable service minimal-foo.service.
Queued /org/freedesktop/systemd1/job/19693 to call RestartUnit on portable service minimal.service.
```

```
$ sudo portablectl detach --runtime --now /tmp/minimal_1 minimal
(Matching unit files with prefixes 'minimal'.)
Queued /org/freedesktop/systemd1/job/19776 to call StopUnit on portable service minimal-foo.service.
Queued /org/freedesktop/systemd1/job/19777 to call StopUnit on portable service minimal.service.
Removed /run/systemd/system.attached/minimal.service.
Removed /run/systemd/system.attached/minimal.service.d/10-profile.conf.
Removed /run/systemd/system.attached/minimal.service.d/20-portable.conf.
Removed /run/systemd/system.attached/minimal.service.d.
Removed /run/systemd/system.attached/minimal-foo.service.
Removed /run/systemd/system.attached/minimal-foo.service.d/10-profile.conf.
Removed /run/systemd/system.attached/minimal-foo.service.d/20-portable.conf.
Removed /run/systemd/system.attached/minimal-foo.service.d.
Removed /run/portables/minimal_1.
Removed /run/systemd/system.attached.
```

keszybz commented on src/portable/portable.c:
```diff
@@ -1035,10 +1035,14 @@ static bool marker_matches_image(const char *marker, const char *name_or_path) {
                 /* We shall match against an image name. In that case let's compare the last component, and optionally
-                 * allow either a suffix of ".raw" or a series of "/". */
+                 * allow either a suffix of ".raw" or a series of "/".
+                 * But allow matching on a differen version of the same image, when a "_" is used as a separator. */
```
typos here.

keszybz commented on src/portable/portable.c:
```diff
@@ -1035,10 +1035,14 @@ static bool marker_matches_image(const char *marker, const char *name_or_path) {
+                 * But allow matching on a differen version of the same image, when a "_" is used as a separator. */
+                underscore = strchr(name_or_path, '_');
+                if (underscore)
+                        return memcmp(a, name_or_path, underscore - name_or_path) == 0;
```
`strneq(a, name_or_path, underscore - name_or_path)`.

Hmm, shouldn't this match be only done for the last component of the path? Matching on `_` in directories seems wrong.

bluca replied:
> Hmm, shouldn't this match be only done for the last component of the path? Matching on _ in directories seems wrong

image_name_is_valid already enforces that name_or_path is not a path, but a single name - ie, without "/" anywhere.

keszybz commented on src/portable/portable.c:
```diff
@@ -1060,6 +1064,10 @@ static bool marker_matches_image(const char *marker, const char *name_or_path) {
+                if (underscore)
+                        l = underscore - b;
+
                 return memcmp(a, b, l) == 0;
```
Here too I guess.

keszybz commented on man/portablectl.xml:
```diff
@@ -155,6 +155,20 @@
+        <listitem><para>Detaches an existing portable service image from the host, and immediately attaches it again.
+        This is useful in case the image was replaced. Running units are not stopped during the process. Partial matching,
+        to allow for different versions in the image name, is allowed: only the part before the first <literal>_</literal>
+        character have to match. If the new image doesn't exist, the existing one will not be detached. The parameters
```
"part … have" → "parts … have" or "part … has"

keszybz commented on man/portablectl.xml:
```diff
@@ -155,6 +155,20 @@
+        follow the same syntax as the <command>attach</command> command.</para></listitem>
+
+        <para>If <option>--now</option> and/or <option>--enable</option> are passed, the portable service(s) are
+        immediately stopped if removed, started and/or enabled if added, or restarted if updated. Prefix(es) are also
```
I'd drop the parens. Just use plural to also mean a single service.

keszybz commented on src/portable/portablectl.c:
```diff
@@ -920,6 +1017,9 @@ static int help(int argc, char *argv[], void *userdata) {
                "                              Detach the specified portable service image\n"
+               "  reattach NAME|PATH [PREFIX...]\n"
+               "                              Reattach and restart the specified portable\n"
+               "                              service image\n"
```
IIUC, "and restart" is only true when `--now` is also given.

keszybz commented on src/portable/portabled-image-bus.c:
```diff
@@ -422,6 +422,190 @@ static int bus_image_method_remove(sd_bus_message *message, void *userdata, sd_b
         return bus_image_common_remove(NULL, message, NULL, userdata, error);
 }
 
+/* Give two PortableChange arrays, return a new array that has all elements of the first that are
```
Give  → Given ?

keszybz commented on src/portable/portabled-image-bus.c:
```diff
@@ -422,6 +422,190 @@ static int bus_image_method_remove(sd_bus_message *message, void *userdata, sd_b
+
+                for (size_t j = 0; j < n_changes_attached; ++j) {
+                        char *attached_name = (char *)basename(changes_detached[i].path),
+                                *detached_name = (char *)basename(changes_attached[j].path);
```
Why is the cast needed here? Can't the type be `const char *`?
Maybe align those variables?
attached/detached seems to be switched on the rhs.
```C
if (streq(basename(changes_detached[i].path), basename(changes_attached[j].path)) {
     found = true;
     break;
}

bluca replied:
Fixed as suggested

keszybz reviewed (changes requested):
Code-wise looks reasonable. I think the semantics are OK, but I'll leave that decision to @poettering.

keszybz commented:
Only minor comments, but I still set the needs-rework flag, because it'd be nice to see those fixed.

bluca commented:
> Only minor comments, but I still set the needs-rework flag, because it'd be nice to see those fixed.

Thanks for the review, updated, PTAL
